clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -3717,6 +3717,16 @@ def err_sme_definition_using_za_in_non_sme_target : Error<
   "function using ZA state requires 'sme'">;
 def err_sme_definition_using_zt0_in_non_sme2_target : Error<
   "function using ZT0 state requires 'sme2'">;
+def warn_sme_streaming_pass_return_vl_to_non_streaming : Warning<
+  "passing a VL-dependent argument to/from a function that has a different"
+  " streaming-mode. The streaming and non-streaming vector lengths may be"
+  " different">,
+  InGroup<AArch64SMEAttributes>, DefaultIgnore;
+def warn_sme_locally_streaming_has_vl_args_returns : Warning<
+  "passing/returning a VL-dependent argument to/from a __arm_locally_streaming"
+  " function. The streaming and non-streaming vector"
+  " lengths may be different">,
+  InGroup<AArch64SMEAttributes>, DefaultIgnore;
```

Comment:
Personally I would prefer it if you could give these warnings a 'bool' operand, such that it would give a message about _either_ the return value or the operand, for example:
```
def warn_sme_streaming_pass_return_vl_to_non_streaming : Warning<
  "%select{passing|returning}0 a VL-dependent argument %select{to|from}0 a function with a different"
  " streaming-mode is undefined behaviour if the streaming and non-streaming vector lengths are different at runtime">,
  InGroup<AArch64SMEAttributes>, DefaultIgnore;
```

And then emit the diagnostic like this:
```
if (ScalableArgs)
  Diag(Loc, diag::warn_sme_streaming_pass_return_vl_to_non_streaming) << /*IsArg=*/true;
if (ScalableRet)
  Diag(Loc, diag::warn_sme_streaming_pass_return_vl_to_non_streaming) << /*IsArg=*/false;
```

This results in:
```
warning: returning a VL-dependent argument from a function with a different streaming-mode is undefined behaviour if the streaming and non-streaming vector lengths are different at runtime
```

and
```
warning: passing a VL-dependent argument to a function with a different streaming-mode is undefined behaviour if the streaming and non-streaming vector lengths are different at runtime
```

I won't hold up the patch for it, but I would appreciate it if you could do this either as part of this PR or as a follow-up patch.